### PR TITLE
optimize watch-cache getlist

### PR DESCRIPTION
**Pull request by sxllwx**
#### What this PR does / why we need it:

Faster watch-cache get-list.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

In order to facilitate verification, I split it into two commits:

- Add benchmarks
- Optimize GetList

> Here are my own benchmark results. FIY...

```bash
# before
$ go test -bench=BenchmarkCacher_GetList  -count 5 . -run=none --benchmem
goos: linux
goarch: amd64
pkg: k8s.io/apiserver/pkg/storage/cacher
cpu: AMD EPYC 7K62 48-Core Processor
BenchmarkCacher_GetList-16    	14	 137743757 ns/op	129016510 B/op	   67975 allocs/op
BenchmarkCacher_GetList-16    	15	 121673849 ns/op	151072125 B/op	   66776 allocs/op
BenchmarkCacher_GetList-16    	15	  88535782 ns/op	151072186 B/op	   66774 allocs/op
BenchmarkCacher_GetList-16    	18	  83706540 ns/op	126228029 B/op	   63987 allocs/op
BenchmarkCacher_GetList-16    	19	 124967934 ns/op	149816737 B/op	   63252 allocs/op
PASS

# after

$ go test -bench=BenchmarkCacher_GetList  . -run=none --benchmem --count=5
goos: linux
goarch: amd64
pkg: k8s.io/apiserver/pkg/storage/cacher
cpu: AMD EPYC 7K62 48-Core Processor
BenchmarkCacher_GetList-16    	      44	  24302640 ns/op	 2528573 B/op	    5749 allocs/op
BenchmarkCacher_GetList-16    	      45	  23997310 ns/op	 2490390 B/op	    5623 allocs/op
BenchmarkCacher_GetList-16    	      46	  23521017 ns/op	 2453680 B/op	    5501 allocs/op
BenchmarkCacher_GetList-16    	      48	  23682351 ns/op	 2384704 B/op	    5273 allocs/op
BenchmarkCacher_GetList-16    	      49	  23772900 ns/op	 2352476 B/op	    5166 allocs/op
PASS

# compare
benchcmp before.text after.text                                                                      
benchmark                      old ns/op     new ns/op     delta
BenchmarkCacher_GetList-16     137743757     24302640      -82.36%
BenchmarkCacher_GetList-16     121673849     23997310      -80.28%
BenchmarkCacher_GetList-16     88535782      23521017      -73.43%
BenchmarkCacher_GetList-16     83706540      23682351      -71.71%
BenchmarkCacher_GetList-16     124967934     23772900      -80.98%

benchmark                      old allocs     new allocs     delta
BenchmarkCacher_GetList-16     67975          5749           -91.54%
BenchmarkCacher_GetList-16     66776          5623           -91.58%
BenchmarkCacher_GetList-16     66774          5501           -91.76%
BenchmarkCacher_GetList-16     63987          5273           -91.76%
BenchmarkCacher_GetList-16     63252          5166           -91.83%

benchmark                      old bytes     new bytes     delta
BenchmarkCacher_GetList-16     129016510     2528573       -98.04%
BenchmarkCacher_GetList-16     151072125     2490390       -98.35%
BenchmarkCacher_GetList-16     151072186     2453680       -98.38%
BenchmarkCacher_GetList-16     126228029     2384704       -98.11%
BenchmarkCacher_GetList-16     149816737     2352476       -98.43%

```

#### Does this PR introduce a user-facing change?

```release-note
Kube-apiserver: Improved memory use when performing GetList on the cache.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by liggitt** [comment]
cc @wojtek-t

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go [hunk 1]
```diff
@@ -1693,3 +1693,59 @@ func TestGetCurrentResourceVersionFromStorage(t *testing.T) {
...
+	cacher, _, err := newTestCacher(newObjectStorage(50000))
+	if err != nil {
+		b.Fatalf("couldn't create cacher: %v", err)
+	}
+	defer cacher.Stop()
+	// prepare result and pred
+	pred := storage.SelectionPredicate{
+		Label: labels.Everything(),
+		Field: fields.Everything(),
+	}
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:1737]: This benchmark assumes that we actually return all objects.

The other extremely important usecase is where a bunch of objects are actually filtered out (e.g. Kubelet listing its own pods).
Can you add a second benchmark that will simulate this one (i.e. only say 50 out of those 50,000 pods are being returned)?

**Comment by sxllwx** [comment]
Update: 

```bash
# before
goos: linux
goarch: amd64
pkg: k8s.io/apiserver/pkg/storage/cacher
cpu: AMD EPYC 7K62 48-Core Processor
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=50-16         	      52	  21252051 ns/op	 1888993 B/op	    4921 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=50-16         	      54	  20719688 ns/op	 1848934 B/op	    4742 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=100-16        	      52	  21634518 ns/op	 3070568 B/op	    5370 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=100-16        	      54	  21208107 ns/op	 2987209 B/op	    5191 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=1000-16       	      51	  22207205 ns/op	 4495174 B/op	    5963 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=1000-16       	      52	  22786041 ns/op	 4425023 B/op	    5869 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=5000-16       	      40	  28037855 ns/op	15776801 B/op	   11318 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=5000-16       	      39	  31645133 ns/op	16156842 B/op	   11476 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=10000-16      	      30	  37621528 ns/op	31225948 B/op	   18409 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=10000-16      	      30	  35822193 ns/op	31225642 B/op	   18408 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=50000-16      	      12	 134383915 ns/op	146229170 B/op	   70962 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=50000-16      	      12	  96080972 ns/op	146228061 B/op	   70955 allocs/op

# after
goos: linux
goarch: amd64
pkg: k8s.io/apiserver/pkg/storage/cacher
cpu: AMD EPYC 7K62 48-Core Processor
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=50-16         	      49	  22527578 ns/op	 1875755 B/op	    5176 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=50-16         	      50	  22053190 ns/op	 1855078 B/op	    5074 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=100-16        	      51	  22239215 ns/op	 2075891 B/op	    4979 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=100-16        	      51	  22236347 ns/op	 2076071 B/op	    4979 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=1000-16       	      51	  22550354 ns/op	 2355291 B/op	    4979 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=1000-16       	      50	  22607157 ns/op	 2375407 B/op	    5078 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=5000-16       	      45	  25300896 ns/op	 4679899 B/op	    5639 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=5000-16       	      44	  24741316 ns/op	 4706111 B/op	    5768 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=10000-16      	      44	  26705112 ns/op	 7613999 B/op	    5769 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=10000-16      	      43	  26508674 ns/op	 7641156 B/op	    5902 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=50000-16      	      30	  34770585 ns/op	31674469 B/op	    8438 allocs/op
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=50000-16      	      31	  33819240 ns/op	31619322 B/op	    8167 allocs/op

# benchcmp

benchmark                                                                  old ns/op     new ns/op     delta
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=50-16        21252051      22527578      +6.00%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=50-16        20719688      22053190      +6.44%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=100-16       21634518      22239215      +2.80%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=100-16       21208107      22236347      +4.85%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=1000-16      22207205      22550354      +1.55%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=1000-16      22786041      22607157      -0.79%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=5000-16      28037855      25300896      -9.76%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=5000-16      31645133      24741316      -21.82%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=10000-16     37621528      26705112      -29.02%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=10000-16     35822193      26508674      -26.00%

benchmark                                                                  old allocs     new allocs     delta
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=50-16        4921           5176           +5.18%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=50-16        4742           5074           +7.00%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=100-16       5370           4979           -7.28%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=100-16       5191           4979           -4.08%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=1000-16      5963           4979           -16.50%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=1000-16      5869           5078           -13.48%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=5000-16      11318          5639           -50.18%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=5000-16      11476          5768           -49.74%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=10000-16     18409          5769           -68.66%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=10000-16     18408          5902           -67.94%

benchmark                                                                  old bytes     new bytes     delta
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=50-16        1888993       1875755       -0.70%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=50-16        1848934       1855078       +0.33%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=100-16       3070568       2075891       -32.39%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=100-16       2987209       2076071       -30.50%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=1000-16      4495174       2355291       -47.60%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=1000-16      4425023       2375407       -46.32%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=5000-16      15776801      4679899       -70.34%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=5000-16      16156842      4706111       -70.87%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=10000-16     31225948      7613999       -75.62%
BenchmarkCacher_GetList/totalObjectNum=50000,_expectObjectNum=10000-16     31225642      7641156       -75.53%
```

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go (hunk 1 above)
Comment (sxllwx) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:1787]: Your prompt reply is much appreciated 😄!  Use-cases have been added and I provided the latest benchmark results at https://github.com/kubernetes/kubernetes/pull/116327#issuecomment-1459733316. PTAL

Comment (sxllwx) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:1787]: Fixed, PTAL @wojtek-t

**Comment by sxllwx** [comment]
@wojtek-t PTAL, thx.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go [hunk 2]
```diff
@@ -1715,3 +1727,111 @@ func TestWaitUntilWatchCacheFreshAndForceAllEvents(t *testing.T) {
...
+}
+
+func BenchmarkCacher_GetList(b *testing.B) {
+	testCases := []struct {
+		name            string
+		totalObjectNum  int
+		expectObjectNum int
+	}{
+		{
+			name:            "totalObjectNum=5000, expectObjectNum=50",
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:1759]: nit: you could construct this name via fmt.Sprintf when calling t.Run, and then there would be zero chance it wouldn't match the test case settings.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 3]
```diff
@@ -773,24 +773,26 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 		return err
 	}
 	span.AddEvent("Listed items from cache", attribute.Int("count", len(objs)))
-	if len(objs) > listVal.Cap() && pred.Label.Empty() && pred.Field.Empty() {
-		// Resize the slice appropriately, since we already know that none
-		// of the elements will be filtered out.
-		listVal.Set(reflect.MakeSlice(reflect.SliceOf(c.objectType.Elem()), 0, len(objs)))
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:779]: would the benchmarks be even better if you just omitted this and continued to use append?

Comment (sxllwx) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:779]: As suggested, I used the following code for testing: [https://github.com/sxllwx/kubernetes/blob/optimize/watch-cache-getlist-fix-comment/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go](https://github.com/sxllwx/kubernetes/blob/optimize/watch-cache-getlist-fix-comment/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go)

The test results are as follows:

```go
# cat cache.before                                                                                                              
goos: linux
goarch: amd64
pkg: k8s.io/apiserver/pkg/storage/cacher
cpu: AMD EPYC 7K62 48-Core Processor
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16         	     847	   1374085 ns/op	  156705 B/op	     132 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16         	     865	   1386726 ns/op	  156078 B/op	     130 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16        	     757	   1513711 ns/op	  626450 B/op	     588 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16        	     750	   1511499 ns/op	  626509 B/op	     588 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16       	     650	   1761892 ns/op	 1696343 B/op	    1095 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16       	     643	   1759985 ns/op	 1696426 B/op	    1096 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16       	     424	   2731813 ns/op	 5431142 B/op	    2619 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16       	     421	   2737815 ns/op	 5431237 B/op	    2619 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16       	     290	   4055868 ns/op	10330039 B/op	    5148 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16       	     286	   4007450 ns/op	10330295 B/op	    5149 allocs/op
PASS
ok  	k8s.io/apiserver/pkg/storage/cacher	34.838s

# cat cache.after
goos: linux
goarch: amd64
pkg: k8s.io/apiserver/pkg/storage/cacher
cpu: AMD EPYC 7K62 48-Core Processor
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16         	     850	   1360398 ns/op	  119860 B/op	      86 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16         	     872	   1364373 ns/op	  119206 B/op	      84 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16        	     817	   1430057 ns/op	  360647 B/op	      89 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16        	     806	   1431489 ns/op	  360690 B/op	      90 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16       	     748	   1491980 ns/op	  640999 B/op	      94 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16       	     747	   1484896 ns/op	  641051 B/op	      94 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16       	     663	   1703111 ns/op	 1456964 B/op	     100 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16       	     673	   1686134 ns/op	 1456867 B/op	     100 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16       	     568	   2003787 ns/op	 2834520 B/op	     109 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16       	     564	   2010633 ns/op	 2834591 B/op	     109 allocs/op
PASS
ok  	k8s.io/apiserver/pkg/storage/cacher	34.016s

# cat cache.comment                                                                                                            
goos: linux
goarch: amd64
pkg: k8s.io/apiserver/pkg/storage/cacher
cpu: AMD EPYC 7K62 48-Core Processor
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16         	     841	   1373561 ns/op	  156746 B/op	     132 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16         	     862	   1366341 ns/op	  156134 B/op	     130 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16        	     769	   1519386 ns/op	  626438 B/op	     587 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16        	     735	   1508256 ns/op	  626657 B/op	     589 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16       	     638	   1764863 ns/op	 1696476 B/op	    1096 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16       	     644	   1759715 ns/op	 1696403 B/op	    1096 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16       	     418	   2721730 ns/op	 5431312 B/op	    2620 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16       	     418	   2732375 ns/op	 5431319 B/op	    2620 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16       	     292	   4007642 ns/op	10329922 B/op	    5148 allocs/op
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16       	     291	   4042856 ns/op	10329993 B/op	    5148 allocs/op
PASS
ok  	k8s.io/apiserver/pkg/storage/cacher	34.775s

# benchcmp cache.before cache.after                                                                                                 
benchcmp is deprecated in favor of benchstat: https://pkg.go.dev/golang.org/x/perf/cmd/benchstat
benchmark                                                                old ns/op     new ns/op     delta
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       1374085       1360398       -1.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       1386726       1364373       -1.61%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      1513711       1430057       -5.53%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      1511499       1431489       -5.29%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1761892       1491980       -15.32%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1759985       1484896       -15.63%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2731813       1703111       -37.66%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2737815       1686134       -38.41%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     4055868       2003787       -50.60%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     4007450       2010633       -49.83%

benchmark                                                                old allocs     new allocs     delta
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       132            86             -34.85%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       130            84             -35.38%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      588            89             -84.86%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      588            90             -84.69%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1095           94             -91.42%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1096           94             -91.42%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2619           100            -96.18%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2619           100            -96.18%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     5148           109            -97.88%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     5149           109            -97.88%

benchmark                                                                old bytes     new bytes     delta
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       156705        119860        -23.51%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       156078        119206        -23.62%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      626450        360647        -42.43%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      626509        360690        -42.43%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1696343       640999        -62.21%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1696426       641051        -62.21%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     5431142       1456964       -73.17%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     5431237       1456867       -73.18%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     10330039      2834520       -72.56%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     10330295      2834591       -72.56%

# benchcmp cache.before cache.comment                                                                                                
benchcmp is deprecated in favor of benchstat: https://pkg.go.dev/golang.org/x/perf/cmd/benchstat
benchmark                                                                old ns/op     new ns/op     delta
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       1374085       1373561       -0.04%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       1386726       1366341       -1.47%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      1513711       1519386       +0.37%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      1511499       1508256       -0.21%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1761892       1764863       +0.17%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1759985       1759715       -0.02%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2731813       2721730       -0.37%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2737815       2732375       -0.20%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     4055868       4007642       -1.19%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     4007450       4042856       +0.88%

benchmark                                                                old allocs     new allocs     delta
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       132            132            +0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       130            130            +0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      588            587            -0.17%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      588            589            +0.17%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1095           1096           +0.09%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1096           1096           +0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2619           2620           +0.04%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     2619           2620           +0.04%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     5148           5148           +0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     5149           5148           -0.02%

benchmark                                                                old bytes     new bytes     delta
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       156705        156746        +0.03%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=50-16       156078        156134        +0.04%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      626450        626438        -0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=500-16      626509        626657        +0.02%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1696343       1696476       +0.01%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=1000-16     1696426       1696403       -0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     5431142       5431312       +0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=2500-16     5431237       5431319       +0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     10330039      10329922      -0.00%
BenchmarkCacher_GetList/totalObjectNum=5000,_expectObjectNum=5000-16     10330295      10329993      -0.00%
```

Conclusion: no noticeable performance gain. (I guess: Golang does Copy internally during the Append process)

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go (hunk 2 above)
Comment (sxllwx) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go]: Thank you for your suggestion. already fixed.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 4]
```diff
@@ -773,24 +773,26 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
...
+			filteredObjects = append(filteredObjects, elem.Object)
 		}
 	}
-	if listVal.IsNil() {
+	if len(filteredObjects) == 0 {
 		// Ensure that we never return a nil Items pointer in the result for consistency.
 		listVal.Set(reflect.MakeSlice(listVal.Type(), 0, 0))
+	} else {
+		// Resize the slice appropriately, since we already know that size of result set
+		listVal.Set(reflect.MakeSlice(listVal.Type(), len(filteredObjects), len(filteredObjects)))
```
[the remark this answers is not in the document]
Comment (sxllwx) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: fixed, PTAL, thx

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 3 above)
Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:779]: Thank you for checking!

**Comment by lavalamp** [comment]
/lgtm
/approve

Thank you! (this should merge when we open up for 1.28)

**Comment by sxllwx** [comment]
> /lgtm
> /approve
> 
> Thank you! (this should merge when we open up for 1.28)

Ok，Thank you for your help, it has been a pleasure to work with you.

**Comment by wojtek-t** [comment]
Sorry for delay - this LGTM too.

/lgtm
/label tide/merge-method-merge

**Comment by sxllwx** [comment]
I want to add a release-note here. hope it's not too late

**Pull request merged**